### PR TITLE
Authorization

### PR DESCRIPTION
**Pull request by stewsk**
Introduce 3 authorization levels:
* read-only : authenticated-user
* run actions : role *reviewer*
* all : role *processor*

nodejs local:  
Add test users with respective roles in package.json (-> switch to authorization strategy "basic-auth" to get login screen)

java local:
Add test users with respective roles in application.yaml. Can be used instead of "admin/admin" in `ui5.yaml`.

Little fix in Java custom handler so that role "reviewer" is sufficient to execute action "Deduct Discount"

### CODE REVIEW

**Comment by stewsk** [comment]
Hi @danjoa or @johannes-vogel ,
either something isn't working as expected or I am doing something wrong - could any of you please have a look?

User "alice" has no role assigned -> my assumption is: can only read.
When running locally (cds w) and logging in with "alice", I can do everything (run actions, delete entries ...)

Now, if I comment out the two lines
https://github.com/SAP-samples/cap-sflight/blob/916d0faeff46498a6ae29a4ae38343ee2f848734/srv/travel-service.cds#L14-L15

in the service definition, it works as expected: "alice" can only read.
But - in my understanding - these two lines should have no effect on "alice".
-> ???

**Comment by stewsk** [comment]
Remark: in the current state the test failures are expected.

**Comment by johannes-vogel** [comment]
> Hi @danjoa or @johannes-vogel , either something isn't working as expected or I am doing something wrong - could any of you please have a look?
> 
> User "alice" has no role assigned -> my assumption is: can only read. When running locally (cds w) and logging in with "alice", I can do everything (run actions, delete entries ...)
> 
> Now, if I comment out the two lines
> 
> https://github.com/SAP-samples/cap-sflight/blob/916d0faeff46498a6ae29a4ae38343ee2f848734/srv/travel-service.cds#L14-L15
> 
> in the service definition, it works as expected: "alice" can only read. But - in my understanding - these two lines should have no effect on "alice". -> ???

alice has the role `admin` by default. Can you try with an empty roles array?

**Comment by stewsk** [comment]
Hi @johannes-vogel ,
yes, that works, thanks.
So I just made an unfortunate choice of my example users :-) - wasn't aware of these default users

**Comment by stewsk** [comment]
Hi @rjayasinghe ,
there is a little mismatch between nodejs and java concerning authorization. In Java:
testuser "martha" has role "reviewer", which allows to execute actions. "accept travel" and "decline travel" works fine, but when executing "deduct discount", there is an error 
```
Not authorized to send event 'UPDATE' to 'TravelService.Travel'
```
On nodejs side there is no such error, i.e. role "reviewer" can perform action "deduct discount".

Is this a general diff between java and node backends, or is this due to the implementation of the custom actions (and thus can be fixed in sflight)?

**Comment by rjayasinghe** [comment]
> Hi @rjayasinghe , there is a little mismatch between nodejs and java concerning authorization. In Java: testuser "martha" has role "reviewer", which allows to execute actions. "accept travel" and "decline travel" works fine, but when executing "deduct discount", there is an error
> 
> ```
> Not authorized to send event 'UPDATE' to 'TravelService.Travel'
> ```
> 
> On nodejs side there is no such error, i.e. role "reviewer" can perform action "deduct discount".
> 
> Is this a general diff between java and node backends, or is this due to the implementation of the custom actions (and thus can be fixed in sflight)?

In Java the role 'reviewer' can perform the actions but it seems that it has no permissions to UPDATE entities. It's debatable whether an UPDATE triggered from a permitted action is also permitted. I'll check with the CAP Java team.

**Comment by stewsk** [comment]
> In Java the role 'reviewer' can perform the actions but it seems that it has no permissions to UPDATE entities. It's debatable whether an UPDATE triggered from a permitted action is also permitted.

My opinion is: if I am authorized to perform an action, then I should not need additional authorization for really being able to use the action.

And: I can perform the "acceptTravel" and "declineTravel", which also perform an UPDATE on Travel -> they change the status

**Comment by rjayasinghe** [comment]
> And: I can perform the "acceptTravel" and "declineTravel", which also perform an UPDATE on Travel -> they change the status

That's true. Let's see what the experts say. :)

**Comment by rjayasinghe** [comment]
@stewsk as always, the differences are subtle: In CAP Java the authorization for DB access is checked only when the statement is accessed on service level (via an instance of `ApplicationService`). If the access happens on DB level (via an instance of `PersistenceService`) no additional authorization checks are performed. In our case the DeductDiscountHandler always uses the the `DraftService` as it needs to be draft-aware. The `DraftService` is also an `ApplicationService`. So with this call
https://github.com/SAP-samples/cap-sflight/blob/c1fc4b5c77f9b370aac9344018f3b4286d1ed164/srv/src/main/java/com/sap/cap/sflight/processor/DeductDiscountHandler.java#L56 the calling user needs UPDATE privileges on the entity Travel.

In https://github.com/SAP-samples/cap-sflight/blob/main/srv/src/main/java/com/sap/cap/sflight/processor/AcceptRejectHandler.java#L81-L90
you can see that in case of draft-enabled handling we use a `DraftService` instance for the UPDATE and in the non-draft case we use a `PersistenceService` that does not perform authorization checks. TBH I cannot tell anymore why I did not choose an `ApplicationService` instance for the non-draft case. This would have been more consistent.

In summary: at least from the CAP Java perspective it would be correct if the `reviewer` role had also UPDATE privileges on the Travel entity. Alternatively, the statement could also be executed in the context of a more privileged user.

@johannes-vogel is there such a differentiation (regarding authorization in DB and service layer) in CAP node.js?

**Comment by johannes-vogel** [comment]
> @johannes-vogel is there such a differentiation (regarding authorization in DB and service layer) in CAP node.js?

Yes, that differentiation is also done in node.js.
The update is done on DB layer without authorization checks:
https://github.com/SAP-samples/cap-sflight/blob/c1fc4b5c77f9b370aac9344018f3b4286d1ed164/srv/travel-service.js#L130-L136

**Comment by stewsk** [comment]
But if role "reviewer" had also UPDATE auth for Travel, he could change all sorts of stuff, that is not intended. I also want him to be able to perform the action.
Is there a way to write the handler in such a way that no additional UPDATE auth is needed?

**Comment by stewsk** [comment]
> In that case we'd need to define a suitable request context that would allow an UPDATE on the Travel entity. I could try to contribute the needed changes to this PR.

That would be cool

**Comment by rjayasinghe** [comment]
@stewsk I pushed a changed version of the custom handlers for deduct discount and accept/reject but I could not really validate if it works now. Please have a look.

**Comment by rjayasinghe** [comment]
@stewsk I now worked with the fixed credentials for 'martha' and I was able to call decuctDiscount and accept/rejectTravel without issues. Issue seems to be fixed, right?

**Comment by chgeo** [comment]
Deployment works as well 👍, see URL above
The [deploy script](https://github.com/SAP-samples/cap-sflight/blob/ef9e31c9b6a67725d583e31ddec3dd60edb08d57/.github/workflows/node.js.yml#L45) assigned all of us the `admin` role, in case you wonder.

File: package.json [hunk 1]
```diff
@@ -40,7 +40,22 @@
       },
       "auth": {
         "[development]": {
-          "strategy": "dummy"
+          "kind": "dummy-auth",
+          "xxx-kind": "basic-auth",
```
Comment (chgeo) [inline package.json:44]: Leftover?

Comment (stewsk) [reply package.json:49]: yes

**Review by chgeo — APPROVED** [review]

**Review by rjayasinghe — APPROVED** [review]
I love it already!
